walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,179 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
+  private int meetingDuration;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // The total (common) free time we have to conduct the meeting.
+    LinkedList<TimeRange> freeTime = new LinkedList<TimeRange>();
+    freeTime.addAll(Arrays.asList(TimeRange.WHOLE_DAY));
+
+    meetingDuration = (int) request.getDuration();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+
+    // If duration of the meeting is greater than a whole day, then return an empty list.
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return freeTime;
+    }
+
+    // To store all the events which the optional attendees are attending.
+    LinkedList<Event> optionalAttendeeEvents = new LinkedList<>();
+
+    /* Changes freeTime such that it now conatins all the free time slots when the meeting can
+    happen, updating the events that optional attendees can attend. */
+    freeSlotsWithMandatoryAttendees(
+        freeTime, events, optionalAttendeeEvents, optionalAttendees, mandatoryAttendees);
+
+    /* If free time is not empty and optional attendees are not empty
+    then go ahead and look for time slots to have meeting with our optional attendees */
+    if (!freeTime.isEmpty() && !optionalAttendeeEvents.isEmpty()) {
+      freeTime =
+          freeSlotsWithOptionalAttendees(optionalAttendeeEvents, freeTime, optionalAttendees);
+    }
+
+    return freeTime;
+  }
+
+  // Loop through all the events given and modify freeTime List
+  private void freeSlotsWithMandatoryAttendees(
+      LinkedList<TimeRange> freeTime,
+      Collection<Event> events,
+      LinkedList<Event> optionalAttendeeEvents,
```

Comment:
These optional attendee parameters are not used since this method only deals with "mandatory attendees", right?

Reply:
So, in this method we loop through all the events and the first 'if' collects all the events that optional attendees are attending thus updating the optionalAttendeeEvents list.

Reply:
Which we will loop through later in the freeSlotsWithOptionalAttendees method.

Reply:
I understand what you mean, I'll remove the optional attendees part from freeSlotsWithMandatoryAttendees and add it to freeSlotsWithOptionalAttendees.

Reply:
And I also got rid of the optionalAttendeeEvents as we don't really require it.